Rust/src/random_cut_forest.rs

```diff
@@ -90,6 +92,60 @@ impl<T> RandomCutForest<T>
         }
     }
 
+    /// Returns the anomaly score associated with the input point relative to
+    /// the data used to update the random cut forest model.
+    ///
+    /// Each constituent random cut tree reports an anomaly score based on its
+    /// random sample of observed data. The forest anomaly score is the mean
+    /// of these per-tree scores.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use random_cut_forest::{RandomCutForest, RandomCutForestBuilder};
+    ///
+    /// // create a default RCF on two-dimensional data
+    /// let mut rcf: RandomCutForest<f32> = RandomCutForestBuilder::new(2).build();
+    ///
+    /// // train the forest on some vector data
+    /// # let data: Vec<Vec<f32>> = vec![vec![0.0, 0.0], vec![1.0, 1.0]];
+    /// for point in data.iter() {
+    ///     rcf.update(point.clone());
+    /// }
+    ///
+    /// // compute an anomaly score on some query point
+    /// let query = vec![0.5, 0.5];
+    /// let score = rcf.anomaly_score(&query);
+    ///
+    /// // compute the anomaly scores of the training points
+    /// let scores: Vec<f32> = data.iter().map(|p| rcf.anomaly_score(p)).collect();
+    /// ```
+    pub fn anomaly_score(&self, point: &Vec<T>) -> T {
+        let anomaly_score: T = self.trees.iter()
+            .map(|t| self.tree_anomaly_score(point, t.tree()))
+            .sum();
+        anomaly_score / T::from(self.num_trees()).unwrap()
+    }
+
+    /// Compute the anomaly score from a particular tree. Used by
+    /// ['Self::anomaly_score'].
+    fn tree_anomaly_score(&self, point: &Vec<T>, tree: &Tree<T>) -> T {
```

Comment:
What's the rationale for putting this here as opposed to the tree implementation?

Reply:
Great question. When I wrote this I put it here without much thought. Of course, it could go into `anomaly_score(), itself. Just wanted to break it out.

That said, there's a bigger problem here: for each "score"/visitor there is an operation on each of the trees and then some kind of combination logic. In the Java version, `traverseForest()` accepts the point along with a visitor (factory), accumulator, and finisher. The former is the visitor, as usual. The latter two are the post-processing.

As we add more scoring algorithms, do we want to add a corresponding function to *both* `Forest` and `Tree`? I detect a code smell in this case. The `traverseForest()` approach seems cleaner, but I wonder if a `ScoringAlgorithm` type would be more appropriate. The `ScoringAlgorithm` (name pending) would include each of the things used by `traverseForest()` plus more. 

At this point, it may be worth forking this discussion out to a tracking issue. I think it's an important problem because I want the addition of new scoring algorithms to be as easy as possible to implement for the developer.

Reply:
Actually, #206 already exists. Will include some of this discussion, there.

Reply:
Right, so far we've been able to get by with just generic traversal methods defined on the tree class. Methods like "get anomaly score" are just UX conveniences for invoking the visitor pattern plus aggregation. Adding new algorithms shouldn't require adding new methods to the tree class.

Reply:
> Adding new algorithms shouldn't require adding new methods to the tree class.

Agreed.